File: server/src/main/java/de/ingrid/igeserver/api/DatasetsApiController.java
package de.ingrid.igeserver.api;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.orientechnologies.orient.core.db.ODatabaseSession;
import de.ingrid.igeserver.db.DBApi;
import de.ingrid.igeserver.db.DBFindAllResults;
import de.ingrid.igeserver.db.FindOptions;
import de.ingrid.igeserver.db.QueryType;
import de.ingrid.igeserver.documenttypes.AddressWrapperType;
import de.ingrid.igeserver.documenttypes.DocumentType;
import de.ingrid.igeserver.model.Data1;
import de.ingrid.igeserver.model.SearchResult;
import de.ingrid.igeserver.services.DocumentService;
import de.ingrid.igeserver.services.MapperService;
import de.ingrid.igeserver.utils.AuthUtils;
import de.ingrid.igeserver.utils.DBUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.security.Principal;
import java.time.OffsetDateTime;
import java.util.*;
import java.util.stream.Collectors;

import static de.ingrid.igeserver.db.OrientDBDatabase.DB_ID;
import static de.ingrid.igeserver.documenttypes.DocumentWrapperType.TYPE;
import static de.ingrid.igeserver.services.MapperService.*;


@RestController
@RequestMapping(path = "/api")
public class DatasetsApiController implements DatasetsApi {

    private static final Logger log = LogManager.getLogger(DatasetsApiController.class);

    private enum CopyMoveOperation {COPY, MOVE}

    private final DBApi dbService;

    private final DocumentService documentService;

    private final DBUtils dbUtils;

    private final AuthUtils authUtils;

    @Autowired
    public DatasetsApiController(AuthUtils authUtils, DBUtils dbUtils, DBApi dbService, DocumentService documentService) {

        this.authUtils = authUtils;
        this.dbUtils = dbUtils;
        this.dbService = dbService;
        this.documentService = documentService;
    }

    /**
     * Create dataset.
     */
    public ResponseEntity<JsonNode> createDataset(
            Principal principal,
            String data,
            boolean address,
            Boolean publish) throws ApiException {

        String userId = this.authUtils.getUsernameFromPrincipal(principal);
        String dbId = this.dbUtils.getCurrentCatalogForUser(userId);

        try (ODatabaseSession ignored = dbService.acquire(dbId)) {

            ObjectNode dataJson = (ObjectNode) getJsonNode(data);

            // add generated id to document (_id)
            // this one is different from the internal database id (@rid)
            UUID uuid = UUID.randomUUID();
            dataJson.put(FIELD_ID, uuid.toString());
            dataJson.put(FIELD_HAS_CHILDREN, false);
            String now = OffsetDateTime.now().toString();
            dataJson.put(FIELD_CREATED, now);
            dataJson.put(FIELD_MODIFIED, now);

            // get document type from document
            String documentType = dataJson.get(FIELD_DOCUMENT_TYPE).asText();

            JsonNode result = this.dbService.save(documentType, null, dataJson.toString());


            JsonNode nodeParentId = dataJson.get(PARENT_ID);
            String parentId = nodeParentId == null ? null : nodeParentId.textValue();

            // create DocumentWrapper
            ObjectNode documentWrapper = this.documentService.getDocumentWrapper();
            documentWrapper.put(FIELD_ID, uuid.toString());
            documentWrapper.put(FIELD_DRAFT, result.get(DB_ID).asText());
            documentWrapper.put(FIELD_PARENT, parentId);

            JsonNode resultWrapper = this.dbService.save(
                    address ? AddressWrapperType.TYPE : TYPE, null, documentWrapper.toString());

            ObjectNode resultDoc = this.documentService.getLatestDocument(resultWrapper);
            return ResponseEntity.ok(resultDoc);
        } catch (Exception e) {
            log.error("Error during creation of document", e);
            throw new ApiException(e.getMessage());
        }

    }

    /**
     * Update dataset.
     */
    public ResponseEntity<JsonNode> updateDataset(
            Principal principal, String id,
            String data,
            boolean publish,
            boolean revert, boolean forAddress) throws ApiException {

        String userId = this.authUtils.getUsernameFromPrincipal(principal);
        String dbId = this.dbUtils.getCurrentCatalogForUser(userId);
        String type = forAddress ? AddressWrapperType.TYPE : TYPE;

        try (ODatabaseSession ignored = dbService.acquire(dbId)) {

            if (dbId == null) {
                throw new NotFoundException(HttpStatus.NOT_FOUND.value(), "The user does not seem to be assigned to any database.");
            }

            if (revert) {
                throw new ApiException("Not implemented");
                // prepareDocumentFromDB = this.jsonFromService.revertDocument( id );
                // return;
            }

            ObjectNode docWrapper = (ObjectNode) this.documentService.getByDocId(id, type, false);


            String recordId;
            // just update document by using new data and adding database ID
            if (docWrapper.get(FIELD_DRAFT).isNull()) {
                // create copy of published document with a new db-id
//                    ObjectNode published = (ObjectNode) docWrapper.get(FIELD_PUBLISHED);
//                    published.remove(DB_ID);
//                    Map newDraft = this.dbService.save(docType, null, published);
                recordId = null; // (String) newDraft.get(DB_ID);
            } else {
                recordId = docWrapper.get(FIELD_DRAFT).asText();
            }

            // save document with same ID or new one, if no draft version exists
            ObjectNode updatedDocument = (ObjectNode) getJsonNode(data);
            updatedDocument.put(FIELD_MODIFIED, OffsetDateTime.now().toString());
            String docType = updatedDocument.get(FIELD_DOCUMENT_TYPE).asText();

            handleLinkedDocs(updatedDocument);

            JsonNode save = this.dbService.save(docType, recordId, updatedDocument.toString());
            String dbID = save.get(DB_ID).asText();

            if (publish) {
                handlePublishingOnWrapper(type, docWrapper, dbID);
            } else {
                handleSaveOnWrapper(type, docWrapper, dbID);
            }

            JsonNode savedDoc = documentService.getByDocId(id, type, true);
            ObjectNode result = documentService.getLatestDocument(savedDoc);

            return ResponseEntity.ok(result);

        } catch (Exception e) {
            log.error("Error during updating of document", e);
            throw new ApiException(e.getMessage());
        }

    }

    private void handleLinkedDocs(ObjectNode doc) throws Exception {

        String docType = doc.get(FIELD_DOCUMENT_TYPE).asText();
        DocumentType refType = documentService.getDocumentType(docType);

        refType.handleLinkedFields(doc, dbService);

    }

    private void handleSaveOnWrapper(String type, ObjectNode docWrapper, String dbID) throws ApiException {
        // update document wrapper with new draft version
        if (docWrapper.get(FIELD_DRAFT).isNull()) {
            // TODO: db_id is ORecord!
            docWrapper.put(FIELD_DRAFT, dbID);
            this.dbService.save(type, docWrapper.get(DB_ID).asText(), docWrapper.toString());
        }
    }

    private void handlePublishingOnWrapper(String type, ObjectNode docWrapper, String dbID) throws ApiException {
        // add ID from published field to archive
        if (!docWrapper.get(FIELD_PUBLISHED).isNull()) {
            docWrapper.withArray(FIELD_ARCHIVE).add(docWrapper.get(FIELD_PUBLISHED));
        }

        // add doc to published reference
        docWrapper.put(FIELD_PUBLISHED, dbID);

        // remove draft version
        docWrapper.put(FIELD_DRAFT, (String) null);

        this.dbService.save(type, docWrapper.get(DB_ID).asText(), docWrapper.toString());
    }

    public ResponseEntity<String> deleteById(Principal principal, String[] ids, boolean forAddress) throws Exception {

        String userId = this.authUtils.getUsernameFromPrincipal(principal);
        String dbId = this.dbUtils.getCurrentCatalogForUser(userId);
        String type = forAddress ? AddressWrapperType.TYPE : TYPE;

        try (ODatabaseSession ignored = dbService.acquire(dbId)) {
            for (String id : ids) {

                // TODO: remove references to document!?
                // TODO: remove all children recursively

                // String recordId = this.dbService.getRecordId(type, id);
                // JsonNode dbDoc = this.dbService.find(type, recordId);

                this.dbService.remove(type, id);

            }
            return ResponseEntity.ok().build();
        }
    }

    public ResponseEntity<Void> copyDatasets(
            Principal principal,
            List<String> ids,
            Data1 data) {

        try {

            copyOrMove(CopyMoveOperation.COPY, ids, data.getDestId());
            return new ResponseEntity<>(HttpStatus.OK);

        } catch (Exception ex) {
            log.error("Error during copy", ex);
            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

    public ResponseEntity<Void> moveDatasets(
            Principal principal,
            List<String> ids,
            Data1 data) throws Exception {

        String userId = this.authUtils.getUsernameFromPrincipal(principal);
        String dbId = this.dbUtils.getCurrentCatalogForUser(userId);

        try (ODatabaseSession ignored = dbService.acquire(dbId)) {

            copyOrMove(CopyMoveOperation.MOVE, ids, data.getDestId());
            return new ResponseEntity<>(HttpStatus.OK);

        }
    }

    private void copyOrMove(CopyMoveOperation operation, List<String> ids, String destId) throws Exception {

        for (String id : ids) {
            JsonNode doc = this.dbService.find(TYPE, id);

            // add new parent to document
            ObjectNode updatedDoc = (ObjectNode) documentService.updateParent(DBUtils.toJsonString(doc), destId);

            if (operation == CopyMoveOperation.COPY) {
                // remove internal dataset Info (TODO: this should be done by the dbService)
                removeDBManagementFields(updatedDoc);

                // when we copy the node, then we also have to reset the id
                updatedDoc.set(MapperService.FIELD_ID, null);
            }

            // TODO: which ID?
            // null should be fine since a new document is created when copied
            // when moved however it should have the same ID!
            this.dbService.save("Documents", null, updatedDoc.toString());

        }
    }

    public ResponseEntity<List<ObjectNode>> getChildren(
            Principal principal,
            String parentId,
            boolean isAddress
    ) throws ApiException {

        DBFindAllResults docs;

        String userId = this.authUtils.getUsernameFromPrincipal(principal);
        String dbId = this.dbUtils.getCurrentCatalogForUser(userId);
        String type = isAddress ? AddressWrapperType.TYPE : TYPE;

        try (ODatabaseSession ignored = dbService.acquire(dbId)) {

            Map<String, String> queryMap = new HashMap<>();
            queryMap.put("_parent", parentId);
            FindOptions findOptions = new FindOptions();
            findOptions.queryType = QueryType.exact;
            findOptions.resolveReferences = true;
            docs = this.dbService.findAll(type, queryMap, findOptions);

            List<ObjectNode> childDocs = docs.hits.stream()
                    .map(doc -> {
                        ObjectNode node = documentService.getLatestDocument(doc);
                        node.put(FIELD_HAS_CHILDREN, this.documentService.determineHasChildren(doc, type));
                        return node;
                    })
                    .collect(Collectors.toList());

            return ResponseEntity.ok(childDocs);
        } catch (Exception e) {
            log.error(e);
            throw new ApiException("Error occured getting children: " + e.getMessage());
        }
    }

    public ResponseEntity<SearchResult> find(Principal principal, String query, int size, String sort, String sortOrder, boolean forAddress) throws Exception {

        DBFindAllResults docs;

        String userId = this.authUtils.getUsernameFromPrincipal(principal);
        String dbId = this.dbUtils.getCurrentCatalogForUser(userId);
        String type = forAddress ? AddressWrapperType.TYPE : TYPE;

        try (ODatabaseSession ignored = dbService.acquire(dbId)) {
            Map<String, String> queryMap = new HashMap<>();
            queryMap.put("draft.title", query);
            queryMap.put("draft IS NULL AND published.title", query);
            FindOptions findOptions = new FindOptions();
            findOptions.size = size;
            findOptions.queryType = QueryType.like;
            findOptions.sortField = sort;
            findOptions.sortOrder = sortOrder;
            findOptions.resolveReferences = true;

            docs = this.dbService.findAll(type, queryMap, findOptions);

            SearchResult searchResult = new SearchResult();
            searchResult.totalHits = docs.totalHits;

            searchResult.hits = docs.hits.stream()
                    .map(documentService::getLatestDocument)
                    .collect(Collectors.toList());

            return ResponseEntity.ok(searchResult);

        }
    }

    public ResponseEntity<JsonNode> getByID(
            Principal principal,
            String id,
            Boolean publish, boolean address) throws Exception {

        String type = address ? AddressWrapperType.TYPE : TYPE;

        String userId = this.authUtils.getUsernameFromPrincipal(principal);
        String dbId = this.dbUtils.getCurrentCatalogForUser(userId);

        try (ODatabaseSession ignored = dbService.acquire(dbId)) {
            Map<String, String> query = new HashMap<>();
            query.put("_id", id);
            FindOptions findOptions = new FindOptions();
            findOptions.queryType = QueryType.exact;
            findOptions.resolveReferences = true;
            DBFindAllResults docs = this.dbService.findAll(type, query, findOptions);

            if (docs.totalHits > 0) {
                ObjectNode doc = documentService.getLatestDocument(docs.hits.get(0));

                return ResponseEntity.ok(doc);
            } else {
                throw new NotFoundException(404, "Document not found with id: " + id);
            }
        }


    }

    public ResponseEntity<List<String>> getPath(
            Principal principal,
            String id,
            boolean forAddress) throws ApiException {

        String userId = this.authUtils.getUsernameFromPrincipal(principal);
        String dbId = this.dbUtils.getCurrentCatalogForUser(userId);
        String type = forAddress ? AddressWrapperType.TYPE : TYPE;

        String destId = id;
        List<String> path = new ArrayList<>();
        path.add(id);

        try (ODatabaseSession ignored = dbService.acquire(dbId)) {
            while (destId != null) {
                JsonNode doc = this.documentService.getByDocId(destId, type, false);
                destId = doc.get("_parent").textValue();
                path.add(destId);
            }
        } catch (Exception e) {
            log.error(e);
        }
        // remove last element which is null
        path.remove(path.size() - 1);

        // turn path around
        Collections.reverse(path);

        return ResponseEntity.ok(path);
    }

}

File: server/src/main/java/de/ingrid/igeserver/api/DatasetsApiController.kt
/**
 * ==================================================
 * Copyright (C) 2023-2024 wemove digital solutions GmbH
 * ==================================================
 * Licensed under the EUPL, Version 1.2 or – as soon they will be
 * approved by the European Commission - subsequent versions of the
 * EUPL (the "Licence");
 *
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */
package de.ingrid.igeserver.api

import com.fasterxml.jackson.databind.JsonNode
import de.ingrid.igeserver.annotations.AuditLog
import de.ingrid.igeserver.model.CopyOptions
import de.ingrid.igeserver.model.DocumentWithMetadata
import de.ingrid.igeserver.model.User
import de.ingrid.igeserver.persistence.FindAllResults
import de.ingrid.igeserver.persistence.postgresql.jpa.model.ige.Document
import de.ingrid.igeserver.persistence.postgresql.jpa.model.ige.DocumentWrapper
import de.ingrid.igeserver.persistence.postgresql.jpa.model.ige.Group
import de.ingrid.igeserver.repository.DocumentRepository
import de.ingrid.igeserver.repository.DocumentWrapperRepository
import de.ingrid.igeserver.services.*
import de.ingrid.igeserver.utils.AuthUtils
import de.ingrid.igeserver.utils.convertToDocument
import de.ingrid.igeserver.utils.prepareDocumentWithMetadata
import de.ingrid.mdek.upload.storage.Storage
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.acls.domain.BasePermission
import org.springframework.security.acls.domain.SidRetrievalStrategyImpl
import org.springframework.security.acls.model.SidRetrievalStrategy
import org.springframework.security.core.Authentication
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.security.Principal
import java.util.*

@RestController
@RequestMapping(path = ["/api"])
class DatasetsApiController(
    private val authUtils: AuthUtils,
    private val catalogService: CatalogService,
    private val docWrapperRepo: DocumentWrapperRepository,
    private val docRepo: DocumentRepository,
    private val documentService: DocumentService,
    private val groupService: GroupService,
    private val aclService: IgeAclService,
    private val storage: Storage,
) : DatasetsApi {

//    private val log = logger()

    private val sidRetrievalStrategy: SidRetrievalStrategy = SidRetrievalStrategyImpl()

    /**
     * Create dataset.
     */
    override fun createDataset(
        principal: Principal,
        data: JsonNode,
        type: String,
        parentId: Int?,
        address: Boolean,
        publish: Boolean
    ): ResponseEntity<DocumentWithMetadata> {
        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
        val doc = convertToDocument(data, type)
        val resultDoc = documentService.createDocument(principal, catalogId, doc, parentId, address, publish)
        val metadata = prepareDocumentWithMetadata(resultDoc)
        return ResponseEntity.ok(metadata)
    }

    /**
     * Update dataset.
     */
    override fun updateDataset(
        principal: Principal,
        id: Int,
        data: JsonNode,
        publishDate: Date?,
        publish: Boolean,
        unpublish: Boolean,
        cancelPendingPublishing: Boolean,
        revert: Boolean,
        version: Int?
    ): ResponseEntity<DocumentWithMetadata> {

        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
        val resultDoc = if (revert) {
            documentService.revertDocument(principal, catalogId, id)
        } else if (unpublish) {
            documentService.unpublishDocument(principal, catalogId, id)
        } else if (cancelPendingPublishing) {
            documentService.cancelPendingPublishing(principal, catalogId, id)
        } else if (publish) {
            val doc = convertToDocument(data, docVersion = version)
            documentService.publishDocument(principal, catalogId, id, doc, publishDate)
        } else {
            val doc = convertToDocument(data, docVersion = version)
            documentService.updateDocument(principal, catalogId, id, doc)
        }

        val metadata = prepareDocumentWithMetadata(resultDoc)
        return ResponseEntity.ok(metadata)
    }


    @Transactional
    override fun deleteById(principal: Principal, ids: List<Int>): ResponseEntity<Unit> {

        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
        for (id in ids) {
            // TODO: remove references to document!?
            documentService.deleteDocument(principal, catalogId, id)
        }
        return ResponseEntity.noContent().build()
    }

    @AuditLog(action = "copy_datasets", target = "options", data = "ids")
    @Transactional
    override fun copyDatasets(
        principal: Principal,
        ids: List<Int>,
        options: CopyOptions
    ): ResponseEntity<List<DocumentWithMetadata>> {
        val destCanWrite = aclService.getPermissionInfo(principal as Authentication, options.destId)
            .let { it.canWrite || it.canOnlyWriteSubtree }
        val sourceCanRead = ids.map { aclService.getPermissionInfo(principal, it).canRead }.all { it }
        if (!(destCanWrite && sourceCanRead)) throw ForbiddenException.withAccessRights("No access to referenced datasets")


        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
        val results = ids.map { id -> handleCopy(principal, catalogId, id, options) }
        return ResponseEntity.ok(results)
    }

    @AuditLog(action = "move_datasets", target = "options", data = "ids")
    @Transactional
    override fun moveDatasets(
        principal: Principal,
        ids: List<Int>,
        options: CopyOptions
    ): ResponseEntity<Void> {
        val destCanWrite = aclService.getPermissionInfo(principal as Authentication, options.destId)
            .let { it.canWrite || it.canOnlyWriteSubtree }
        val sourceCanWrite = ids.map { aclService.getPermissionInfo(principal, it).canWrite }.all { it }
        if (!(destCanWrite && sourceCanWrite)) throw ForbiddenException.withAccessRights("No access to referenced datasets")

        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
        ids.forEach { id -> handleMove(catalogId, id, options) }
        return ResponseEntity(HttpStatus.OK)
    }

    @Transactional
    override fun setResponsibleUser(principal: Principal, datasetId: Int, userId: Int): ResponseEntity<Void> {
        this.docWrapperRepo.findById(datasetId).ifPresent { wrapper ->
            val user = this.catalogService.getUser(userId)
            if (user != null) {
                wrapper.responsibleUser = user
                docWrapperRepo.save(wrapper)
            }
        }
        return ResponseEntity(HttpStatus.OK)
    }

    override fun replaceAddress(
        principal: Principal,
        source: String,
        target: String
    ): ResponseEntity<Unit> {
        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
        this.documentService.replaceAddress(catalogId, source, target)
        return ResponseEntity(HttpStatus.OK)
    }

    override fun setTags(principal: Principal, id: Int, tags: TagRequest): ResponseEntity<List<String>> {
        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
        val updatedTags = this.documentService.updateTags(catalogId, id, tags) ?: emptyList()
        return ResponseEntity.ok(updatedTags)
    }

    /*
     * Validation errors are thrown as exceptions
     */
    override fun validate(principal: Principal, id: Int): ResponseEntity<Unit> {
        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
        documentService.validate(principal, catalogId, id)
        return ResponseEntity.ok().build()
    }

    private fun handleCopy(
        principal: Principal,
        catalogId: String,
        id: Int,
        options: CopyOptions
    ): DocumentWithMetadata {

        val wrapper = documentService.getWrapperByDocumentIdAndCatalog(id)

        val doc = documentService.getDocumentByWrapperId(catalogId, id)

        if (options.includeTree) {
            validateCopyOperation(catalogId, id, options.destId)
        }

//        val metadata = prepareDocumentWithMetadata(DocumentData(wrapper, doc))
        // TODO AW: is there another way?
//        metadata.document.put(FIELD_PARENT, options.destId)

        // clear UUID to create a new one during copy
        documentService.detachDocumentFromDatabase(doc)
        doc.uuid = ""
        doc.id = null
        doc.version = null

        return createCopyAndHandleSubTree(principal, catalogId, doc, options, documentService.isAddress(wrapper))
    }

    private fun createCopyAndHandleSubTree(
        principal: Principal,
        catalogId: String,
        doc: Document,
        options: CopyOptions,
        isAddress: Boolean
    ): DocumentWithMetadata {
        val origParentId = doc.wrapperId!! // doc[FIELD_ID].asInt()
        val origParentUUID = doc.uuid // doc[FIELD_UUID].asText()

        val copiedParent =
            documentService.createDocument(
                principal,
                catalogId,
                doc,
                options.destId,
                isAddress,
                false,
                InitiatorAction.COPY
            )


        storage.copyToUnpublished(catalogId, origParentUUID, copiedParent.wrapper.uuid)

//        val copiedParentJson = documentService.convertToJsonNode(copiedParent.document) as ObjectNode
        if (options.includeTree) {
            val count = handleCopySubTree(principal, catalogId, copiedParent.wrapper.id!!, origParentId, isAddress)
            // TODO AW: is this needed to calc hasChildren?
//            copiedParentJson.put(FIELD_HAS_CHILDREN, count > 0)
        }

        return prepareDocumentWithMetadata(copiedParent)
    }

    private fun handleCopySubTree(
        principal: Principal,
        catalogId: String,
        parentId: Int,
        origParentId: Int,
        isAddress: Boolean
    ): Long {

        // get all children of parent and save those recursively
        val docs = documentService.findChildrenDocs(catalogId, origParentId, isAddress)

        docs.hits.forEach { child ->

            child.let {
//                val childDoc = documentService.getLatestDocument(it, false, false)
//                val childVersion = (documentService.convertToJsonNode(it.document) as ObjectNode)
//                    .put(FIELD_PARENT, parentId)

                // clear UUID to create a new one during copy
                it.document.uuid = ""
                createCopyAndHandleSubTree(principal, catalogId, it.document, CopyOptions(parentId, true), isAddress)
            }

        }
        return docs.totalHits
    }

    private fun handleMove(catalogId: String, id: Int, options: CopyOptions) {

        if (id == options.destId) {
            throw ConflictException.withMoveConflict("Cannot move '$id' to itself")
        }
        validateCopyOperation(catalogId, id, options.destId)

        documentService.updateParent(catalogId, id, options.destId)
    }

    private fun validateCopyOperation(catalogId: String, sourceId: Int, destinationId: Int?) {
        // check destination is not part of source
        val descIds = documentService.getAllDescendantIds(catalogId, sourceId)
        if (descIds.contains(destinationId) || sourceId == destinationId) {
            throw ConflictException.withCopyConflict("Cannot copy '$sourceId' since  '$destinationId' is part of the hierarchy")
        }
    }

    override fun getChildren(
        principal: Principal,
        parentId: String?,
        isAddress: Boolean
    ): ResponseEntity<List<DocumentInfo>> {

        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
        val isSuperOrCatAdmin = authUtils.isAdmin(principal)
        val hasRootRead = checkForRootPermissions(
            sidRetrievalStrategy.getSids(principal as Authentication),
            listOf(BasePermission.READ)
        )

        val childrenInfo = if (
            parentId == null && !isSuperOrCatAdmin && !hasRootRead
        ) {
            // Calculate Root Objects for non-admin users
            val userName = authUtils.getUsernameFromPrincipal(principal)
            val userGroups = catalogService.getUser(userName)?.getGroupsForCatalog(catalogId)
            getRootDocsFromGroup(userGroups, isAddress)
        } else {
            documentService.findChildrenDocs(catalogId, parentId?.toInt(), isAddress)
        }

        val childDocs = childrenInfo.hits
            .map { mapToDocumentInfo(it, isAddress) }
        return ResponseEntity.ok(childDocs)
    }

    private fun mapToDocumentInfo(data: DocumentData, isAddress: Boolean): DocumentInfo {
        return DocumentInfo(
            data.wrapper.id!!,
            data.document.title ?: "???",
            data.document.uuid,
            data.document.state.getState(),
            data.wrapper.countChildren > 0,
            data.wrapper.parent?.id,
            data.wrapper.type,
            data.document.modified!!,
            data.document.contentmodified!!,
            data.wrapper.pending_date,
            data.wrapper.tags.joinToString(","),
            data.wrapper.hasWritePermission,
            data.wrapper.hasOnlySubtreeWritePermission,
            isAddress
        )
    }

    private fun getRootDocsFromGroup(
        userGroups: Set<Group>?,
        isAddress: Boolean,
    ): FindAllResults<DocumentData> {

        val actualRoots = mutableListOf<DocumentWrapper>()
        val groupDatasets = aclService.getDatasetIdsFromGroups(userGroups!!, isAddress)
            .map { id -> documentService.getWrapperById(id) }
        val groupDatasetIds = groupDatasets.map { it.id }.toSet()

        for (potentialRoot in groupDatasets) {
            // if potentialRoot.path contains(intersects) any other groupDatasetId it is a descendant and therefore not an actual root node.
            val isRoot = potentialRoot.path.intersect(groupDatasetIds).isEmpty()
            if (isRoot) actualRoots.add(potentialRoot)
        }

        return documentService.getDocumentsFromWrappers(actualRoots)
    }


    data class UserAccessResponse(val canOnlyRead: Set<User>, val canWrite: Set<User>)

    override fun getUsersWithAccessToDocument(
        principal: Principal,
        id: Int,
    ): ResponseEntity<UserAccessResponse> {
        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
        val isSuperOrCatAdmin = authUtils.isAdmin(principal)

        val canWriteUsers =
            groupService.getUsersWithAccess(principal, catalogId, id, BasePermission.WRITE).toMutableSet()
        val canOnlyReadUsers =
            groupService.getUsersWithAccess(principal, catalogId, id, BasePermission.READ).minus(canWriteUsers)
                .toMutableSet()


        if (isSuperOrCatAdmin) {
            //  only show admins to other admins
            val admins =
                catalogService.getAllCatalogUsers(principal, catalogId).filter { user -> user.role == "cat-admin" }
                    .toSet()
            // admins can always write
            canWriteUsers.addAll(admins)
            canOnlyReadUsers.removeAll(admins)
        } else {
            val principalManagedUsers = catalogService.getAllCatalogUserIds(principal)
                .filter { catalogService.canEditUser(principal, it) }
            val principalLogin = authUtils.getUsernameFromPrincipal(principal)
            // filter list to only show users which the md-admin principal can edit + himself
            canWriteUsers.retainAll { principalManagedUsers.contains(it.login) || it.login == principalLogin }
            canOnlyReadUsers.retainAll { principalManagedUsers.contains(it.login) || it.login == principalLogin }
        }

        return ResponseEntity.ok(UserAccessResponse(canOnlyReadUsers, canWriteUsers))
    }

    override fun getByUUID(
        principal: Principal,
        uuid: String,
        publish: Boolean?
    ): ResponseEntity<DocumentWithMetadata> {
        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
        val wrapper = documentService.getWrapperByCatalogAndDocumentUuid(catalogId, uuid)

        return if (publish == true) {
            val document = documentService.getLastPublishedDocument(catalogId, uuid)
            val metadata = prepareDocumentWithMetadata(DocumentData(wrapper, document))
            ResponseEntity.ok(metadata)
        } else {
            getByID(principal, wrapper.id!!)
        }
    }

    override fun getByID(
        principal: Principal,
        id: Int
    ): ResponseEntity<DocumentWithMetadata> {

        try {
            val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
            // TODO: catalogId is not necessary anymore
            val docData = documentService.getDocumentFromCatalog(catalogId, id, true)
            val metadata = prepareDocumentWithMetadata(docData)
            return ResponseEntity.ok(metadata)
        } catch (ex: AccessDeniedException) {
            throw ForbiddenException.withAccessRights("No read access to document")
        }

    }

    override fun getPath(
        principal: Principal,
        id: Int
    ): ResponseEntity<List<PathResponse>> {
        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)
        val documentData = documentService.getDocumentFromCatalog(catalogId, id)
        val path = documentData.wrapper.path

        val pathInfos = path
            .map { pathId -> createPathResponse(pathId, principal) }

        return ResponseEntity.ok(
            pathInfos + PathResponse(
                documentData.wrapper.id!!,
                documentData.document.title ?: "???",
                PermissionInfo(
                    true,
                    documentData.wrapper.hasWritePermission,
                    documentData.wrapper.hasOnlySubtreeWritePermission
                )
            )
        )

    }

    private fun createPathResponse(
        pathId: Int,
        principal: Principal
    ): PathResponse {
        // use function without permission check to get parents we don't have access to
        val pathWrapper = docWrapperRepo.findByIdNoPermissionCheck(pathId)
        val latestDoc = docRepo.getByCatalogAndUuidAndIsLatestIsTrue(pathWrapper.catalog!!, pathWrapper.uuid)
        val title = latestDoc.title ?: "???"
        val permission = aclService.getPermissionInfo(principal as Authentication, pathId)
        return PathResponse(pathId, title, permission)
    }

    data class PathResponse(val id: Int, val title: String, val permission: PermissionInfo? = null)

}
